beacon_node/network/src/subnet_service/mod.rs

```diff
@@ -17,12 +32,640 @@ pub enum SubnetServiceMessage {
+                    // peer before they can be removed.
+                    warn!(self.log,
+                        "Not enough time for a discovery search";
+                        "subnet_id" => ?subnet,
+                    );
+                    None
+                }
+            })
+            .collect();
+
+        if !discovery_subnets.is_empty() {
+            self.queue_event(SubnetServiceMessage::DiscoverPeers(discovery_subnets));
+        }
+        Ok(())
+    }
+
+    // Subscribes to the subnet if it should be done immediately, or schedules it if required.
+    fn subscribe_to_subnet(
+        &mut self,
+        ExactSubnet { subnet, slot }: ExactSubnet,
+    ) -> Result<(), &'static str> {
+        // If the subnet is one of our permanent subnets, we do not need to subscribe.
+        if self.subscribe_all_subnets || self.permanent_attestation_subscriptions.contains(&subnet)
+        {
+            return Ok(());
+        }
+
+        let slot_duration = self.beacon_chain.slot_clock.slot_duration();
+
+        // The short time we schedule the subscription before it's actually required. This
+        // ensures we are subscribed on time, and allows consecutive subscriptions to the same
+        // subnet to overlap, reducing subnet churn.
+        let advance_subscription_duration = slot_duration / ADVANCE_SUBSCRIBE_SLOT_FRACTION;
+        // The time to the required slot.
+        let time_to_subscription_slot = self
+            .beacon_chain
+            .slot_clock
+            .duration_to_slot(slot)
+            .unwrap_or_default(); // If this is a past slot we will just get a 0 duration.
+
+        // Calculate how long before we need to subscribe to the subnet.
+        let time_to_subscription_start =
+            time_to_subscription_slot.saturating_sub(advance_subscription_duration);
+
+        // The time after a duty slot where we no longer need it in the `aggregate_validators_on_subnet`
+        // delay map.
+        let time_to_unsubscribe =
+            time_to_subscription_slot + UNSUBSCRIBE_AFTER_AGGREGATOR_DUTY * slot_duration;
+        if let Some(tracked_vals) = self.aggregate_validators_on_subnet.as_mut() {
+            tracked_vals.insert_at(ExactSubnet { subnet, slot }, time_to_unsubscribe);
+        }
+
+        // If the subscription should be done in the future, schedule it. Otherwise subscribe
+        // immediately.
+        if time_to_subscription_start.is_zero() {
+            // This is a current or past slot, we subscribe immediately.
+            self.subscribe_to_subnet_immediately(subnet, slot + 1)?;
+        } else {
+            // This is a future slot, schedule subscribing.
+            self.scheduled_subscriptions
+                .insert_at(subnet, time_to_subscription_start);
+        }
+
+        Ok(())
+    }
+
+    /// Adds a subscription event and an associated unsubscription event if required.
```

Comment:
This fn no longer adds unsubscription event:

```suggestion
    /// Adds a sync subscription event with an expiry specified by `duration_to_unsubscribe`.
    /// Once the duration elapses, the subscription will be unsubscribed automatically.
```
```